server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerAdsWarmingClusterIT.java

```diff
@@ -0,0 +1,492 @@
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.Listener;
+import io.envoyproxy.envoy.api.v2.RouteConfiguration;
+import io.envoyproxy.envoy.api.v2.core.AggregatedConfigSource;
+import io.envoyproxy.envoy.api.v2.core.ConfigSource;
+import io.envoyproxy.envoy.api.v2.core.Http2ProtocolOptions;
+import io.grpc.netty.NettyServerBuilder;
+import io.restassured.http.ContentType;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.Network;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
+import org.testcontainers.shaded.org.apache.commons.lang.math.RandomUtils;
+
+public class DiscoveryServerAdsWarmingClusterIT {
+
+  private static final String CONFIG = "envoy/ads.config.yaml";
+  private static final String GROUP = "key";
+  private static final Integer LISTENER_PORT = 10000;
+  private static final SimpleCache<String> cache = new SimpleCache<>(node -> GROUP);
+
+  private static final CountDownLatch onStreamOpenLatch = new CountDownLatch(1);
+  private static final CountDownLatch onStreamRequestLatch = new CountDownLatch(1);
+  private static final CountDownLatch onStreamResponseLatch = new CountDownLatch(1);
+
+  private static final NettyGrpcServerRule ADS = new NettyGrpcServerRule() {
+    @Override
+    protected void configureServerBuilder(NettyServerBuilder builder) {
+      ExecutorService executorService = Executors.newSingleThreadExecutor();
+      final DiscoveryServerCallbacks callbacks = new DiscoveryServerCallbacks() {
+        @Override
+        public void onStreamOpen(long streamId, String typeUrl) {
+          onStreamOpenLatch.countDown();
+        }
+
+        @Override
+        public void onStreamRequest(long streamId, DiscoveryRequest request) {
+          onStreamRequestLatch.countDown();
+        }
+
+        @Override
+        public void onStreamResponse(long streamId, DiscoveryRequest request, DiscoveryResponse response) {
+          if (request.getTypeUrl().equals(Resources.CLUSTER_TYPE_URL)) {
+            executorService.submit(() -> cache.setSnapshot(
+                GROUP,
```

Comment:
Maybe extract this to method `createSnapshotWithWorkingClusterWithTheSameEDSVersion`, just like `createSnapshotWithNotWorkingCluster`, to make it easier to understand?

Also maybe it is worth to add a comment explaining what's going on. That we update a Snapshot with working cluster, but we change only CDS version, not EDS version, to test if EDS will be sent anyway after CDS was sent. It took me a while to understand this and I still don't know if I understand it correctly :)

Reply:
Added description and changed method name